apps/meteor/app/e2e/server/functions/fetchUsersWaitingForGroupKey.ts

```diff
@@ -0,0 +1,46 @@
+import type { IRoom, IUser } from '@rocket.chat/core-typings';
+import { Rooms, Subscriptions, Users } from '@rocket.chat/models';
+
+import { isTruthy } from '../../../../lib/isTruthy';
+
+const NUMBER_OF_ROOMS = 10;
+const QUEUE_SIZE = 50;
+
+export async function fetchUsersWaitingForGroupKey(userId: IUser['_id']): Promise<{
+	usersWaitingForE2EKeys: Record<IRoom['_id'], { _id: IUser['_id']; public_key: string }[]>;
+	hasMore: boolean;
+}> {
+	const roomIds = (await Subscriptions.findByUserId(userId, { projection: { rid: 1 } }).toArray()).map((sub) => sub.rid);
+
+	const cursor = Rooms.findRoomsByRoomIdsWithE2EEQueue(roomIds, { projection: { usersWaitingForE2EKeys: 1 } });
+
+	const count = await cursor.count();
+
+	const hasMore = count > NUMBER_OF_ROOMS;
+	const encryptedRoomIds = (await cursor.toArray()).map((room) => room._id);
+
+	const rooms = await Rooms.fetchRandomUsersFromE2EEQueue(encryptedRoomIds, NUMBER_OF_ROOMS, QUEUE_SIZE);
```

Comment:
We can't rely on the users in the queue only, what if more than 50 users are waiting for the keys?

We must retrieve the users based on the subscriptions (the ones that don't contain E2EKey (`findByRidWithoutE2EKey`)), get all the user ids, and then, from those users, find all users WITH the public key in place, so finally from those users we can extract a `$sample` of users.